python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -21,131 +21,95 @@
+from ._lookupStaticCalibration import lookupStaticCalibration
+
+
+class BrighterFatterKernelSolveConnections(pipeBase.PipelineTaskConnections,
+                                           dimensions=("instrument", "exposure", "detector")):
+    dummy = cT.Input(
+        name="raw",
+        doc="Dummy exposure.",
+        storageClass='Exposure',
+        dimensions=("instrument", "exposure", "detector"),
+        multiple=True,
+        deferLoad=True,
+    )
+    camera = cT.PrerequisiteInput(
+        name="camera",
+        doc="Camera associated with this data.",
+        storageClass="Camera",
+        dimensions=("instrument", ),
+        isCalibration=True,
+        lookupFunction=lookupStaticCalibration,
+    )
+    inputPtc = cT.PrerequisiteInput(
+        name="ptc",
+        doc="Photon transfer curve dataset.",
+        storageClass="PhotonTransferCurveDataset",
+        dimensions=("instrument", "detector"),
+        isCalibration=True,
+    )
 
+    outputBFK = cT.Output(
+        name="brighterFatterKernel",
+        doc="Output measured brighter-fatter kernel.",
+        storageClass="BrighterFatterKernel",
+        dimensions=("instrument", "detector"),
+        isCalibration=True,
+    )
 
-class MakeBrighterFatterKernelTaskConfig(pexConfig.Config):
-    """Config class for bright-fatter effect coefficient calculation."""
 
-    isr = pexConfig.ConfigurableField(
-        target=IsrTask,
-        doc="""Task to perform instrumental signature removal""",
-    )
-    isrMandatorySteps = pexConfig.ListField(
-        dtype=str,
-        doc="isr operations that must be performed for valid results. Raises if any of these are False",
-        default=['doAssembleCcd']
-    )
-    isrForbiddenSteps = pexConfig.ListField(
+class BrighterFatterKernelSolveConfig(pipeBase.PipelineTaskConfig,
+                                      pipelineConnections=BrighterFatterKernelSolveConnections):
+    level = pexConfig.ChoiceField(
+        doc="The level at which to calculate the brighter-fatter kernels",
         dtype=str,
-        doc="isr operations that must NOT be performed for valid results. Raises if any of these are True",
-        default=['doFlat', 'doFringe', 'doBrighterFatter', 'doUseOpticsTransmission',
-                 'doUseFilterTransmission', 'doUseSensorTransmission', 'doUseAtmosphereTransmission']
+        default="AMP",
+        allowed={
+            "AMP": "Every amplifier treated separately",
+            "DETECTOR": "One kernel per detector",
+        }
     )
-    isrDesirableSteps = pexConfig.ListField(
+    ignoreAmpsForAveraging = pexConfig.ListField(
         dtype=str,
-        doc="isr operations that it is advisable to perform, but are not mission-critical."
-        " WARNs are logged for any of these found to be False.",
-        default=['doBias', 'doDark', 'doCrosstalk', 'doDefect', 'doLinearize']
+        doc="List of amp names to ignore when averaging the amplifier kernels into the detector"
```

Comment:
If I remember properly, you mentioned that you got better results when averaging over all the amps compared to when you calculate individual kernels per amp. Was this due to the quality of the input data or was there something more fundamental to it? (that should be documented in a comment in the docstring)

Reply:
That was based on imSim data, for which there should be a single known truth.  Comparing `AMP` kernels against this truth resulted in offsets, such that AmpX would have a kernel with a central value -10% compared to the known value, and AmpY would have a central value +10%.  Switching to a detector kernel averages those differences away, yielding a result that seemed accurate to the few percent level.  The exact levels I'm not sure of ("10%" being my best recollection).

That said, I think there will need to be some significant testing and regeneration of brighter-fatter kernels before they should be trusted.  I know that enabling the `ci_cpp_gen3` brighter fatter kernel produces obviously worse results.  As there isn't a real PTC ramp in that data, I think this is just a result of bad inputs making a bad product.

I'm also seeing that this option is unused in the code, which needs to be fixed.  There are also the `badAmps` and `valid` parameters of the kernel that should be used more consistently.  It looks like there's at least one bug in how they're handled in the `ip_isr` ticket (`fromTable` seems to construct a `badAmp` list from the wrong selection of `valid`).